tests/test-runner/bin/test-runner.py

```diff
@@ -106,7 +108,7 @@ def _read(self):
         for easy sorting/merging later.
         """
         fd = self.fileno()
-        buf = os.read(fd, 4096)
+        buf = bytes.decode(os.read(fd, 4096))
```

Comment:
The Coverage builder (Python 3.5) is now dying here (http://build.zfsonlinux.org/builders/Ubuntu%2017.04%20x86_64%20Coverage%20%28TEST%29/builds/4335/steps/shell_9/logs/tests):

```
Traceback (most recent call last):
  File "/var/lib/buildbot/slaves/zfs/Ubuntu_17_04_x86_64_Coverage__TEST_/build/zfs/tests/test-runner/bin/test-runner.py", line 922, in <module>
    main()
  File "/var/lib/buildbot/slaves/zfs/Ubuntu_17_04_x86_64_Coverage__TEST_/build/zfs/tests/test-runner/bin/test-runner.py", line 917, in main
    testrun.run(options)
  File "/var/lib/buildbot/slaves/zfs/Ubuntu_17_04_x86_64_Coverage__TEST_/build/zfs/tests/test-runner/bin/test-runner.py", line 728, in run
    self.testgroups[testgroup].run(self.logger, options)
  File "/var/lib/buildbot/slaves/zfs/Ubuntu_17_04_x86_64_Coverage__TEST_/build/zfs/tests/test-runner/bin/test-runner.py", line 479, in run
    test.run(options)
  File "/var/lib/buildbot/slaves/zfs/Ubuntu_17_04_x86_64_Coverage__TEST_/build/zfs/tests/test-runner/bin/test-runner.py", line 235, in run
    self.result.stdout, self.result.stderr = self.collect_output(proc)
  File "/var/lib/buildbot/slaves/zfs/Ubuntu_17_04_x86_64_Coverage__TEST_/build/zfs/tests/test-runner/bin/test-runner.py", line 201, in collect_output
    fd.read()
  File "/var/lib/buildbot/slaves/zfs/Ubuntu_17_04_x86_64_Coverage__TEST_/build/zfs/tests/test-runner/bin/test-runner.py", line 99, in read
    while self._read() is not None:
  File "/var/lib/buildbot/slaves/zfs/Ubuntu_17_04_x86_64_Coverage__TEST_/build/zfs/tests/test-runner/bin/test-runner.py", line 111, in _read
    buf = bytes.decode(os.read(fd, 4096))
UnicodeDecodeError: 'utf-8' codec can't decode byte 0xe1 in position 687: invalid continuation byte
```

It still troubles me buildbot does not mark the " zfstests" step as "FAILED" in the build summary: does it make sense to abort the build (and report) when such unhandled errors happens?

```diff
diff --git a/scripts/zfs-tests.sh b/scripts/zfs-tests.sh
index 6fdd658f7..996dca699 100755
--- a/scripts/zfs-tests.sh
+++ b/scripts/zfs-tests.sh
@@ -583,8 +583,11 @@ REPORT_FILE=$(mktemp -u -t zts-report.XXXX -p "$FILEDIR")
 #
 msg "${TEST_RUNNER} ${QUIET} -c ${RUNFILE} -T ${TAGS} -i ${STF_SUITE}" \
     "-I ${ITERATIONS}"
+set -o pipefail
 ${TEST_RUNNER} ${QUIET} -c "${RUNFILE}" -T "${TAGS}" -i "${STF_SUITE}" \
     -I "${ITERATIONS}" 2>&1 | tee "$RESULTS_FILE"
+RESULT=$? # <- fail with this != 0
+set +o pipefail
 
 #
 # Analyze the results.
```

Reply:
Right, I definitely think this should fail.  That last thing we want is to incorrectly reporting success here.  Let me also pull in your `pipefail` change.